HelpersWayland.cpp

```diff
@@ -0,0 +1,295 @@
+
+    xcb_window_t window = xcb_generate_id(connection);
+    if (window <= 0)
+        error_fatal("failed to generate X window id");
+
+    xcb_void_cookie_t create_cookie = xcb_create_window_checked(
+        connection,
+        XCB_COPY_FROM_PARENT, // depth
+        window,
+        screen->root, // parent window
+        window_x,
+        window_y,
+        window_width,
+        window_height,
+        0, // border width
+        XCB_WINDOW_CLASS_INPUT_OUTPUT, // class
+        screen->root_visual, // visual
+        xcb_window_attrib_mask,
+        xcb_window_attrib_list);
+
+    xcb_void_cookie_t map_cookie = xcb_map_window_checked(connection, window);
+
+    // Check errors.
+    error = xcb_request_check(connection, create_cookie);
+    if (error)
+        error_fatal("failed to create X window: %d", error->error_code);
+    error = xcb_request_check(connection, map_cookie);
+    if (error)
+        error_fatal("failed to map X window: %d", error->error_code);
+
+    *out_display = display;
+    *out_connection = connection;
+    *out_window = window;
+}
+
+const EGLint egl_config_attribs[] = {
+    EGL_COLOR_BUFFER_TYPE,     EGL_RGB_BUFFER,
+    EGL_BUFFER_SIZE,           32,
+    EGL_RED_SIZE,              8,
+    EGL_GREEN_SIZE,            8,
+    EGL_BLUE_SIZE,             8,
+    EGL_ALPHA_SIZE,            8,
+
+    EGL_DEPTH_SIZE,            24,
+    EGL_STENCIL_SIZE,          8,
+
+    EGL_SAMPLE_BUFFERS,        0,
+    EGL_SAMPLES,               0,
+
+    EGL_SURFACE_TYPE,          EGL_WINDOW_BIT,
+    EGL_RENDERABLE_TYPE,       EGL_OPENGL_BIT,
+
+    EGL_NONE,
+};
+
+const EGLint egl_context_attribs[] = {
+    EGL_NONE,
+};
+
+const EGLint egl_surface_attribs[] = {
+    EGL_RENDER_BUFFER, EGL_BACK_BUFFER,
+    EGL_NONE,
+};
+
+void HelpersWayland::setup_egl(
+        EGLint api,
+        EGLNativeDisplayType native_display,
+        EGLNativeWindowType native_window,
+        EGLDisplay* out_display,
+        EGLConfig* out_config,
+        EGLContext* out_context,
+        EGLSurface* out_window_surface) {
+
+    EGLint ignore;
+    EGLBoolean ok;
+
+    ok = eglBindAPI(api);
+    if (!ok)
+        error_fatal("eglBindAPI(0x%x) failed", api);
+
+    EGLDisplay display = eglGetDisplay(native_display);
+    if (display == EGL_NO_DISPLAY)
+        error_fatal("eglGetDisplay() failed");
+
+    ok = eglInitialize(display, &ignore, &ignore);
+    if (!ok)
+        error_fatal("eglInitialize() failed");
+
+    EGLint configs_size = 256;
+    EGLConfig* configs = new EGLConfig[configs_size];
+    EGLint num_configs;
+    ok = eglChooseConfig(
+        display,
+        egl_config_attribs,
+        configs,
+        configs_size, // num requested configs
+        &num_configs); // num returned configs
+    if (!ok)
+        error_fatal("eglChooseConfig() failed");
+    if (num_configs == 0)
+        error_fatal("failed to find suitable EGLConfig");
+    EGLConfig config = configs[0];
+    delete [] configs;
```

Comment:
I'm not sure, but I think you have a dangling pointer here.
If you have just set the value of the array to other variable and now delete it, you are using a dangling pointer.

Reply:
we need only the first element of `configs` array. That's why we are deleting it after we retrieve it

Reply:
Yes, but you are deleting the first too.
What happens if you try to access any data from the first element after the `delete` statement?

Reply:
We are not accessing any data from it after deleting. 

I think we don't have any problem with it.

Reply:
If you are passing a `undefined behavior variable` to the EGL api, what is the point in doing it after all? The result is absolutely unpredictable. It can crash the EGL library, it can simple consider that it's null and continue... we simple don't know.
You must choose to:
1. Correctly pass the variable to the API
2. Just pass null, as it seems the API doesn't care for this configuration object.

Reply:
If you we me to remove this line, I'll remove.